src/import/raft_applier.rs

```diff
@@ -0,0 +1,444 @@
+// Copyright 2023 TiKV Project Authors. Licensed under Apache-2.0.
+//! This module contains types for asynchronously applying the write batches
+//! into the storage.
+
+use std::{
+    collections::HashMap,
+    sync::{Arc, Mutex, Weak},
+};
+
+use futures::{Stream, StreamExt};
+use kvproto::kvrpcpb::Context;
+use sst_importer::metrics::{APPLIER_ENGINE_REQUEST_DURATION, APPLIER_EVENT, IMPORTER_APPLY_BYTES};
+use tikv_kv::{with_tls_engine, Engine, WriteData, WriteEvent};
+use tikv_util::{fut, time::Instant};
+use tokio::sync::{OwnedSemaphorePermit, Semaphore};
+
+use crate::storage;
+
+async fn wait_write(mut s: impl Stream<Item = WriteEvent> + Send + Unpin) -> storage::Result<()> {
+    match s.next().await {
+        Some(WriteEvent::Finished(Ok(()))) => Ok(()),
+        Some(WriteEvent::Finished(Err(e))) => Err(e.into()),
+        Some(e) => Err(box_err!("unexpected event: {:?}", e)),
+        None => Err(box_err!("stream closed")),
+    }
+}
+
+const MAX_CONCURRENCY_PER_REGION: usize = 16;
+
+async fn acquire_semaphore(smp: &Arc<Semaphore>) -> Option<OwnedSemaphorePermit> {
+    if let Ok(pmt) = Arc::clone(smp).try_acquire_owned() {
+        return Some(pmt);
+    }
+    APPLIER_EVENT.with_label_values(&["raft-throttled"]).inc();
+    Arc::clone(smp).acquire_owned().await.ok()
+}
+
+#[repr(transparent)]
+#[derive(Clone, Default)]
+pub struct ThrottledTlsEngineWriter(Arc<Mutex<Inner>>);
+
+pub struct GcHandle(Weak<Mutex<Inner>>);
```

Comment:
Why weak?

Reply:
Or you would just check the ref count. Since you are already doing it for individual semaphore, it makes sense to reuse the same logic for the bigger part.